fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/RecordCursorTest.java

```diff
@@ -1012,21 +1014,97 @@ public void testFlatMapReasons() throws Exception {
         assertEquals(Collections.emptyList(), cursor.asList().join());
         RecordCursorResult<Integer> noNextResult = cursor.getNext();
         assertEquals(RecordCursor.NoNextReason.TIME_LIMIT_REACHED, noNextResult.getNoNextReason());
-        cursor = new FakeOutOfBandCursor<>(RecordCursor.fromList(list, noNextResult.getContinuation().toBytes()), 3)
-                .filter(i -> false)
-                .orElse(orElse);
+
+        cursor = RecordCursor.orElse(cont -> new FakeOutOfBandCursor<>(RecordCursor.fromList(list, cont), 3)
+                .filter(i -> false), orElse, noNextResult.getContinuation().toBytes());
         assertEquals(Collections.singletonList(0), cursor.asList().join());
         noNextResult = cursor.getNext();
         assertEquals(RecordCursor.NoNextReason.SOURCE_EXHAUSTED, noNextResult.getNoNextReason());
     }
 
+    @Test
+    public void orElseWithEventuallyNonEmptyInner() {
+        final List<Integer> list = Arrays.asList(1, 2, 3, 4, 5);
+        RecordCursor<Integer> cursor = getFilteredFakeOutOfBandCursor(list, 3, 4, null);
+        assertEquals(Collections.emptyList(), cursor.asList().join());
+        RecordCursorResult<Integer> nextResult = cursor.getNext();
+        assertEquals(RecordCursor.NoNextReason.TIME_LIMIT_REACHED, nextResult.getNoNextReason());
+        cursor = getFilteredFakeOutOfBandCursor(list, 3, 4, nextResult.getContinuation().toBytes());
+
+        // Choose the inner branch since we eventually get a value
+        assertEquals(Collections.singletonList(5), cursor.asList().join());
+        nextResult = cursor.getNext();
+        assertEquals(RecordCursor.NoNextReason.SOURCE_EXHAUSTED, nextResult.getNoNextReason());
+    }
+
+    @Test
+    public void orElseContinueWithInnerBranchAfterDecision() {
+        final List<Integer> longList = Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18);
+        RecordCursor<Integer> cursor = getFilteredFakeOutOfBandCursor(longList, 3, 10, null);
+
+        RecordCursorResult<Integer> nextResult = null;
+        for (int i = 0; i < 3; i++) { // three rounds with no results
+            nextResult = cursor.getNext();
+            assertEquals(RecordCursor.NoNextReason.TIME_LIMIT_REACHED, nextResult.getNoNextReason());
+            cursor = getFilteredFakeOutOfBandCursor(longList, 3, 10, nextResult.getContinuation().toBytes());
+        }
+
+        // Choose the inner branch
+        assertEquals(ImmutableList.of(11, 12), cursor.asList().join());
+        nextResult = cursor.getNext();
+        assertEquals(RecordCursor.NoNextReason.TIME_LIMIT_REACHED, nextResult.getNoNextReason());
+
+        byte[] continuation = nextResult.getContinuation().toBytes();
+        while (continuation != null) {
+            cursor = getFilteredFakeOutOfBandCursor(longList, 3, 10, continuation);
+            nextResult = cursor.getNext();
+            continuation = nextResult.getContinuation().toBytes();
```

Comment:
It feels like this loop should assert something, even if it's just that it never gets "0" back (so that we know it never hits the else branch), though having it assert about the real contents would be better.

Reply:
Okay, I changed this to assert about the real contents.

There is an interesting philosophical question here about testing (i.e., should you assert on everything you can think of, or just the bare minimum behavior you are trying to demonstrate). I'm happy to go along, though.

Reply:
I suppose so. It does seem weird to have this loop at all if it's not going to assert on anything (unless the point is just to prove it eventually completes or something)

Reply:
Right, I was trying to demonstrate that the cursor can indeed be restarted without falling over on continuation parsing exceptions (as happened during testing).